### ISSUE TITLE
Enhancement request: Verbatim file/folder structure on disk

### ISSUE THREAD

**Issue by himselfv** [danny0838/firefox-scrapbook#40]
I'm using Scrapbook with sync solutions such as Dropbox and BTSync. They work on file level. If I modify File.A on one PC and File.B on another, they'll be synchronized. If I modify the same file in two places at once then the oldest version wins and I have to go resolve the conflict manually.

This is acceptable since it's rare and when I do modify one file from two places red light is already flashing in my head and I'm quite attentive to what I'm doing.

In Scrapbook, files are also tracked by scrapbook.rdf. So when I make changes to a folder structure from two places at once, only one change is kept, even if they were unrelated.

This, on the other hand, happens often. Running Firefox on two PCs at the same time is nothing unusual, and you can easily forget you moved some folder on one PC and that Firefox instance is still running (changes not commited/propagated/read back on another Firefox instance).

I'm not sure at which point Scrapbook saves / reloads scrapbook.rdf, surprisingly, I haven't encountered the problem all that often. Perhaps I'm careful enough. But the thought bugs me.

So I thought, why does Scrapbook even use one single file to track everything, when it basically just keeps a file/folder structure? That's what files/folders are for!

**Why not just keep a verbatim folder/file structure on disk, just like it looks in Scrapbook?**
#### There are many advantages to using verbatim structure:
1. Simplicity. Removes part of the equation, expressing the functionality in simpler terms.
2. Stability. Scrapbook.rdf is a point of failure, which is evidenced by the need to keep backups and "Restore scrapbook.rdf" function. Moreover, it's an artificial point of failure, as it's most often failure mode (afaik) is "data inconsistent with what is on disk". Just leave only what is on disk!
3. Compatibility. All the tools in the world work with files as units of information and folders as a unit of grouping. Whatever cool enhancements build on that abstraction (such as sync solutions and perhaps others), Scrapbook would automatically benefit.
4. Isolation. Changes are local, only affect that which was changed.
5. Portability. One of the shortcomings of Scrapbook as a note library is that you cannot take it with you on Android/iPhone. But if the structure was stored verbatim, you can just sync the folder to Android and browse/edit with any of the number of tools.
6. Openness. You don't need data in a special format to be used with Scrapbook. Ideally, you can just drop the .txt file into the Scrapbook folder, and it'll display it as a note (although without any Scrapbook-private features).
#### Why could scrapbook.rdf be needed in the first place?
- To distinguish between note/page types? Lots of apps do this using just files. Use a file extension or if you insist on all files being HTML, add some meta tags into the header.
- To store additional data? Store it as meta (or other) tags in HTML. HTML is designed to be extensible, why not use it.
- Data which is too large/unwieldy for meta tags can also be stored in an external annotation file, the important point is that it's kept close to the main file, and that it's optional - if it's lost, the main file is still visible, but perhaps without some customizations.
- To allow grouping of several files into one note? This can be done by straying just a bit from the verbatim structure. Let folders which have index.html inside be treated as files, where Scrapbook is concerned. Or do it the IE way where folders with the same name as HTML are treated as resource folders. (I'd go the latter way as it's a bit more compliant to established practice, but this point is not important)
- Loading speed? But you don't need to load the whole tree immediately as Scrapbook starts. Just scan the root folder, and then scan folders in the background if you care, and force-scan a subfolder when the user expands it.

Since no one did this before, I assume there are some complexities here I'm not thinking about. But please at least consider the idea and if those can after all be solved! To me it seems that verbatim data structure on disk would be of benefit in many ways to advanced users, while changing almost nothing to Scrapbook users who just work through the interface.

Thanks!

**Comment by danny0838** [comment]
Thank you for providing the idea and a thorough analysis. We do are considering a rework of the data structure. One of the main difficulty is to be downward compatible with the old data structure, and we have to at least make an (almost) bidirectional conversion strategy.

Also, it's so far beyond the current missions and therefore it would probably pend until we are developing ScrapBook X 2.0.

**Comment by pbpb** [comment]
I made a similar request in reference to 'output tree as html.' (Issue #245)  Because the latter can be done independently of revising the whole data structure, and does not have to later be input or backward compatible, it could be done sooner, and later provide the starting point for a more complete rewrite of the data structure.

**Comment by danny0838** [comment]
We are not going to do this in ScrapBook X. We may consider implementation of such feature for [Web ScrapBook](https://github.com/danny0838/webscrapbook), though.
